### PR TITLE
feat(build): Integrate unit test framework

### PR DESCRIPTION
**Pull request by kartikCypherock**
Integrate Unity from https://github.com/ThrowTheSwitch/Unity/commit/c40b80a917d77a715ebcb28ed5aea4409f5d5abc.

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [CHI-4597 Integrate unity C testing framework for unit testing](https://app.clickup.com/t/9002019994/CHI-4597)

File: utilities/run_unit_tests.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    echo "Flashing unit tests on the target"
+    echo "STM32_Programmer_CLI -c port=swd -w build/Main/Cypherock-Main.bin 0x08023000"
+    STM32_Programmer_CLI -c port=swd -w build/Main/Cypherock-Main.bin 0x08023000
+
+    echo "Running unit tests on the target"
+    echo "STM32_Programmer_CLI -c port=swd -startswv freq=80 portnumber=all -RA"
+    STM32_Programmer_CLI -c port=swd -startswv freq=80 portnumber=all -RA
+else
+    echo "Running unit tests on the simulator"
+    ./bin/Cypherock_Simulator.exe
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply utilities/run_unit_tests.sh:22]: fixed

File: utilities/run_unit_tests.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+echo "Invoking build script"
+echo "./utilities/build.sh "$1" "unit_tests" "$2""
+
+./utilities/build.sh "$1" "unit_tests" "$2"
+
+if [ "$2" == "device" ]; then
+    echo "Flashing unit tests on the target"
+    echo "STM32_Programmer_CLI -c port=swd -w build/Main/Cypherock-Main.bin 0x08023000"
+    STM32_Programmer_CLI -c port=swd -w build/Main/Cypherock-Main.bin 0x08023000
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply utilities/run_unit_tests.sh:15]: thanks for this catch.. fixed

File: utilities/run_unit_tests.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+#usage: ./utilities/run_unit_tests.sh [main|initial] [device|simulator]
+
+set -e
+
+echo "Invoking build script"
+echo "./utilities/build.sh "$1" "unit_tests" "$2""
```
Comment (amanCypherock) [inline utilities/run_unit_tests.sh:8]: Add a comments if wrong parameters are given to the run_unit_tests script.
The clickup documentation mentions `./utilities/run_tests.sh [device|simulator]` while actually you need to provide 2 arguments.

Comment (kartikCypherock) [reply utilities/run_unit_tests.sh:41]: I have updated the document

File: utilities/cmake/firmware/firmware.cmake [hunk 4]
```diff
@@ -8,6 +8,19 @@ set(CMAKE_C_STANDARD_REQUIRED ON)
 set(CMAKE_C_EXTENSIONS OFF)
 
 file(GLOB_RECURSE SOURCES "stm32-hal/*.*" "common/*.*" "src/*.*")
```
Comment (ujjwal-cyph) [inline utilities/cmake/firmware/firmware.cmake:10]: remove this duplicate line

File: utilities/run_unit_tests.sh [hunk 5]
```diff
@@ -0,0 +1,55 @@
+#!/bin/sh
+
+usage () {
+    echo -e "\tUSAGE: $0 [main|initial] [device|simulator]"
```
Comment (ujjwal-cyph) [inline utilities/run_unit_tests.sh:4]: the parameters are required. we can use angular brackets here.

```suggestion
    echo -e "\tUSAGE: $0 <main|initial> <device|simulator>"
```

Comment (kartikCypherock) [reply utilities/run_unit_tests.sh]: fixed

File: utilities/cmake/firmware/firmware.cmake (hunk 4 above)
Comment (kartikCypherock) [reply utilities/cmake/firmware/firmware.cmake]: fixed

**Review by ujjwal-cyph — APPROVED** [review]

**Pull request merged**